### DISCUSSION TITLE
saving nifti segmentation back to original(before transformation)

### DISCUSSION THREAD

**Discussion by janetkok** [Project-MONAI/MONAI#1397] [Q&A] [answered]
We normally perform several transformation(eg. resized,spacing,orientation) on our inputs to train our model. As such, when we need to predict a new input data, we will also need to transform the data to get the segmentation. However, the segmentation produced is also a transformed data. Hence, when we want to overlay it on top of our original input data, it will not be allign with it.
How can I make it align with my original data(like doing a revert transform?)?

![image](https://user-images.githubusercontent.com/45690357/103161198-196bca00-47d6-11eb-85ee-9c4ad27c604f.png)

Thanks a lot.

**Comment by Nic-Ma** [comment, marked as answer]
Hi @janetkok ,

Thanks for your interest here.
Currently, not all the transforms update the affine matrix, so we can't revert all the transforms, we will work on it soon, tracking: https://github.com/Project-MONAI/MONAI/issues/365
For your case, can you only use `spacingd` and `orientationd` transforms then `save_nifti` with meta_data? refer to:
https://github.com/Project-MONAI/tutorials/blob/master/3d_segmentation/torch/unet_evaluation_dict.py#L96

Thanks.

**Reply by janetkok** [reply]
Hi, thanks for the answer. I have to apply the resized transformation. If not, I will get the following error:
` stack expects each tensor to be equal size, but got [1, 250, 250, 163] at entry 0 and [1, 271, 271, 139] at entry 1`.

Guess we will not have any solution for that at the moment

**Reply by Nic-Ma** [reply]
This is because you want to batch data with different size, could you please change your batch size to `1` during inference?

Thanks.

**Reply by janetkok** [reply]
Hi,

my batch size during inference has always been 1. This error will occur when I try to train the data if I did not resized them before training.
I had to perform a resized transformation due to the varying volume size of the CT dataset (40~123)x512x512 . 

Thanks

**Reply by janetkok** [reply]
@Nic-Ma 
Another thing to point out is that I tried to save the nifti files(both CT scan and ground truth) after the transformation and it turns out to be like this. I lay the mask on top of the CT scan
![image](https://user-images.githubusercontent.com/45690357/103218521-1b877300-4913-11eb-961e-02b9ddcd849d.png)

Before transformation:
![image](https://user-images.githubusercontent.com/45690357/103218348-a74ccf80-4912-11eb-9237-542638108cab.png)


For your reference, my transformation code is as follows:
```
patch_size = [192, 160,56]
spacing = [1.0, 1.0, 1.0]

train_transforms = Compose(
    [
      LoadImaged(keys=["image", "label"]),
      SqueezeDimd(keys=["image", "label"], dim=None),
      AddChanneld(keys=["image", "label"]),
      Spacingd(keys=["image", "label"], pixdim=spacing,
                 mode=('bilinear','nearest') ),
      Orientationd(keys=["image", "label"], axcodes="RAS"),
      Resized(keys=["image", "label"],mode=("trilinear")  ,spatial_size=patch_size),
      NormalizeIntensityd(keys=["image"], nonzero=True, channel_wise=True),
      RandScaleIntensityd(keys=["image"], factors=0.3, prob=0.15),
      ToTensord(keys=["image", "label"]),
    ]
)
val_transforms = Compose(
    [
      LoadImaged(keys=["image", "label"]),
      SqueezeDimd(keys=["image", "label"], dim=None),
      AddChanneld(keys=["image", "label"]),
      Spacingd(keys=["image", "label"], pixdim=spacing,
                 mode=("bilinear", "nearest")),
      Orientationd(keys=["image", "label"], axcodes="RAS"),
      Resized(keys=["image", "label"],mode=("trilinear") ,spatial_size=patch_size),
      NormalizeIntensityd(keys=["image"], nonzero=True, channel_wise=True),
      ToTensord(keys=["image", "label"]),
    ]
)
````

**Reply by janetkok** [reply]
anyway, i just solved the above weird mask overlay problem by changing the resized mode to nearest. :)

**Reply by Nic-Ma** [reply]
Sounds good, `nearest` makes sense for this case.

**Comment by Nic-Ma** [comment]
And for the overlay question, I think you can train with `resize transform`, and do inference without `resize` then overlay to image.

Thanks.
